### PR TITLE
Implement receive deferred messages

### LINKED ISSUE #10113: Implement ReceiveDeferred

### PR DESCRIPTION
**Pull request by JoshLove-msft**
Moved anything involving RequestResponseLockedMessages down to AmqpConsumer.

Fixes https://github.com/Azure/azure-sdk-for-net/issues/10113

### CODE REVIEW

File: sdk/servicebus/Azure.Messaging.ServiceBus/src/Amqp/AmqpConsumer.cs [hunk 1]
```diff
@@ -11,23 +11,25 @@
...
     ///
-    /// <seealso cref="Azure.Messaging.ServiceBus.Core.TransportConsumer" />
+    /// <seealso cref="Azure.Messaging.ServiceBus.Core.TransportReceiver" />
     ///
-    internal class AmqpConsumer : TransportConsumer
+    internal class AmqpReceiver : TransportReceiver
     {
         /// <summary>The default prefetch count to use for the consumer.</summary>
-        private const uint DefaultPrefetchCount = 0;
+        private const int DefaultPrefetchCount = 0;
```
Comment (jsquire) [inline sdk/servicebus/Azure.Messaging.ServiceBus/src/Amqp/AmqpConsumer.cs]: I'd  recommend leaving this as `uint` unless you've got a reason not to.   The track one implementation pushes this around as `int` only to end up converting it when it gets used with the link.   _(see [here](https://github.com/Azure/azure-sdk-for-net/blob/master/sdk/servicebus/Microsoft.Azure.ServiceBus/src/Core/MessageReceiver.cs#L243))_

File: sdk/servicebus/Azure.Messaging.ServiceBus/src/Amqp/AmqpConsumer.cs [hunk 2]
```diff
@@ -55,6 +57,7 @@ internal class AmqpConsumer : TransportConsumer
         ///
         private string PartitionId { get; }
```
Comment (jsquire) [inline sdk/servicebus/Azure.Messaging.ServiceBus/src/Amqp/AmqpConsumer.cs]: Probably want to nix this, as its an Event Hubs concept.

File: sdk/servicebus/Azure.Messaging.ServiceBus/src/Amqp/AmqpConsumer.cs [hunk 3]
```diff
@@ -336,6 +418,486 @@ private void CloseLink(ReceivingAmqpLink link)
...
+                await _retryPolicy.RunOperation(
+                    async (timeout) => await AbandonInternalAsync(message, timeout, propertiesToModify).ConfigureAwait(false),
+                    EntityName,
+                    ConnectionScope,
+                    cancellationToken).ConfigureAwait(false);
+            }
+            catch (Exception exception)
+            {
+                // MessagingEventSource.Log.MessageAbandonException(ClientId, exception);
+                throw exception;
```
Comment (jsquire) [inline sdk/servicebus/Azure.Messaging.ServiceBus/src/Amqp/AmqpConsumer.cs]: Should this just be `throw` to preserve the original state and metadata?

Comment (jsquire) [reply sdk/servicebus/Azure.Messaging.ServiceBus/src/Amqp/AmqpConsumer.cs]: Leaving just the one comment, but the same pattern exists elsewhere in the class.

File: sdk/servicebus/Azure.Messaging.ServiceBus/src/Receiver/ServiceBusReceiverClient.cs [hunk 4]
```diff
@@ -81,240 +81,120 @@ public class ServiceBusReceiverClient : IAsyncDisposable
         ///   and should take responsibility for managing its lifespan.
         /// </summary>
         ///
-        internal bool OwnsConnection { get; set; } = true;
+        internal bool OwnsConnection { get; set; } = false;
```
Comment (jsquire) [inline sdk/servicebus/Azure.Messaging.ServiceBus/src/Receiver/ServiceBusReceiverClient.cs]: Was this change intentional?   I see it getting flipped to `false` for the protected mocking constructor, but I don't see it set `true` elsewhere.

File: sdk/servicebus/Azure.Messaging.ServiceBus/src/Amqp/AmqpConsumer.cs (hunk 1 above)
Comment (JoshLove-msft) [reply sdk/servicebus/Azure.Messaging.ServiceBus/src/Amqp/AmqpConsumer.cs]: This was based on a comment from Krzysztof.

File: sdk/servicebus/Azure.Messaging.ServiceBus/src/Receiver/ServiceBusReceiverClient.cs (hunk 4 above)
Comment (JoshLove-msft) [reply sdk/servicebus/Azure.Messaging.ServiceBus/src/Receiver/ServiceBusReceiverClient.cs]: Yeah based on the Api Updates PR, this will always be false. It can probably be removed entirely.

File: sdk/servicebus/Azure.Messaging.ServiceBus/src/Amqp/AmqpConsumer.cs (hunk 1 above)
Comment (jsquire) [reply sdk/servicebus/Azure.Messaging.ServiceBus/src/Amqp/AmqpConsumer.cs]: I don't feel strongly, but am curious as to the rationale.   It's an internal value within this class, so any public-facing concerns would be moot.  Why not use the native type that gets sent and avoid the cast?

Comment (JoshLove-msft) [reply sdk/servicebus/Azure.Messaging.ServiceBus/src/Amqp/AmqpConsumer.cs]: The public type defined on the Options needs to be int. This can be uint, but we still have to do the cast at some point.

Comment (jsquire) [reply sdk/servicebus/Azure.Messaging.ServiceBus/src/Amqp/AmqpConsumer.cs]: No argument on that, but the cast here would potentially happen multiple times versus a single time when creating this class.   I don't think its a major concern, just seems to be an area for an easy efficiency.

Comment (JoshLove-msft) [reply sdk/servicebus/Azure.Messaging.ServiceBus/src/Amqp/AmqpConsumer.cs]: I guess it would happen multiple times in the case of retries, but there would generally only ever be one call to CreateReceivingLinkAsync per client, so it seemed more straightforward to just keep things consistent in terms of types.

Comment (jsquire) [reply sdk/servicebus/Azure.Messaging.ServiceBus/src/Amqp/AmqpConsumer.cs]: Remember that `CreateReceivingLink` is potentially called multiple times behind the scenes as well, such as when the link/session is abandoned due to inactivity or timeouts or there is a transient issue with network connectivity.   That's handled for you by the `FaultTolerantAmqpObject` which calls in to create the link on your behalf.

Comment (JoshLove-msft) [reply sdk/servicebus/Azure.Messaging.ServiceBus/src/Amqp/AmqpConsumer.cs]: Ok you've convinced me 😀

File: sdk/servicebus/Azure.Messaging.ServiceBus/src/Amqp/AmqpConsumer.cs (hunk 3 above)
Comment (JoshLove-msft) [reply sdk/servicebus/Azure.Messaging.ServiceBus/src/Amqp/AmqpConsumer.cs]: This stuff will be fixed in a separate PR that will do a sweep of all exceptions.

**Review by jsquire — APPROVED** [review]

File: sdk/servicebus/Azure.Messaging.ServiceBus/src/Amqp/AmqpConsumer.cs [hunk 5]
```diff
@@ -336,6 +418,486 @@ private void CloseLink(ReceivingAmqpLink link)
...
+            finally
+            {
+                // diagnosticSource.DisposeStop(activity, lockToken, deferTask?.Status);
+            }
+
+            cancellationToken.ThrowIfCancellationRequested<TaskCanceledException>();
+            // MessagingEventSource.Log.MessageDeferStop(ClientId);
+        }
+
+
```
Comment (ShivangiReja) [inline sdk/servicebus/Azure.Messaging.ServiceBus/src/Amqp/AmqpConsumer.cs]: nit: extra line

File: sdk/servicebus/Azure.Messaging.ServiceBus/src/Amqp/AmqpConsumer.cs [hunk 6]
```diff
@@ -336,6 +418,486 @@ private void CloseLink(ReceivingAmqpLink link)
...
+        }
+
+        /// <summary>
+        /// Abandons a <see cref="ServiceBusMessage"/> using a lock token. This will make the message available again for processing.
+        /// </summary>
+        ///
+        /// <param name="message">The lock token of the corresponding message to abandon.</param>
+        /// <param name="timeout"></param>
+        /// <param name="propertiesToModify">The properties of the message to modify while abandoning the message.</param>
+        internal virtual Task AbandonInternalAsync(
```
Comment (ShivangiReja) [inline sdk/servicebus/Azure.Messaging.ServiceBus/src/Amqp/AmqpConsumer.cs]: nit: We are not appending `async` in other  internal methods(`CompleteInternal`, `DeferInternal`... etc), for the consistency can we update this to `AbandonInternal` or add `async` in other methods?

File: sdk/servicebus/Azure.Messaging.ServiceBus/src/Amqp/AmqpConsumer.cs [hunk 7]
```diff
@@ -336,6 +418,486 @@ private void CloseLink(ReceivingAmqpLink link)
...
+        /// <summary>
+        /// Moves a message to the deadletter sub-queue.
+        /// </summary>
+        ///
+        /// <param name="message">The lock token of the corresponding message to deadletter.</param>
+        /// <param name="timeout"></param>
+        /// <param name="propertiesToModify">The properties of the message to modify while moving to sub-queue.</param>
+        /// <param name="deadLetterReason">The reason for deadlettering the message.</param>
+        /// <param name="deadLetterErrorDescription">The error description for deadlettering the message.</param>
+        internal virtual Task DeadLetterInternalAsync(
```
Comment (ShivangiReja) [inline sdk/servicebus/Azure.Messaging.ServiceBus/src/Amqp/AmqpConsumer.cs]: nit: same as above. For the consistency  can we remove async or update other methods to append async?

File: sdk/servicebus/Azure.Messaging.ServiceBus/src/Amqp/AmqpConsumer.cs [hunk 8]
```diff
@@ -336,6 +418,486 @@ private void CloseLink(ReceivingAmqpLink link)
...
+        /// <param name="sessionId"></param>
+        /// <param name="propertiesToModify"></param>
+        /// <param name="deadLetterReason"></param>
+        /// <param name="deadLetterDescription"></param>
+        private async Task DisposeMessageRequestResponseAsync(
+            Guid[] lockTokens,
+            TimeSpan timeout,
+            DispositionStatus dispositionStatus,
+            bool isSessionReceiver,
+            string sessionId = null,
```
Comment (ShivangiReja) [inline sdk/servicebus/Azure.Messaging.ServiceBus/src/Amqp/AmqpConsumer.cs]: I believe you've removed these params in other PR and you need to pull latest from the master.

File: sdk/servicebus/Azure.Messaging.ServiceBus/src/Amqp/AmqpConsumer.cs [hunk 9]
```diff
@@ -425,46 +1030,108 @@ internal List<ArraySegment<byte>> ConvertLockTokensToDeliveryTags(IEnumerable<Gu
...
-            return lockedUntilUtc;
+        /// <summary>
+        /// Receives a <see cref="IList{Message}"/> of deferred messages identified by <paramref name="sequenceNumbers"/>.
+        /// </summary>
+        /// <param name="sequenceNumbers">An <see cref="IEnumerable{T}"/> containing the sequence numbers to receive.</param>
+        /// <param name="cancellationToken"></param>
+        /// <returns>Messages identified by sequence number are returned. Returns null if no messages are found.
+        /// Throws if the messages have not been deferred.</returns>
+        /// <seealso cref="DeferAsync"/>
+        public override async Task<IList<ServiceBusReceivedMessage>> ReceiveDeferredMessageBatchAsync(
```
Comment (ShivangiReja) [inline sdk/servicebus/Azure.Messaging.ServiceBus/src/Amqp/AmqpConsumer.cs]: `Task<IList<ServiceBusReceivedMessage>>`  -> `Task<IEnumerable<ServiceBusReceivedMessage>>` ?

File: sdk/servicebus/Azure.Messaging.ServiceBus/src/Amqp/AmqpConsumer.cs [hunk 10]
```diff
@@ -425,46 +1030,108 @@ internal List<ArraySegment<byte>> ConvertLockTokensToDeliveryTags(IEnumerable<Gu
...
+            IEnumerable<long> sequenceNumbers,
+            CancellationToken cancellationToken = default)
+        {
+            Argument.AssertNotClosed(IsClosed, nameof(AmqpReceiver));
```
Comment (ShivangiReja) [inline sdk/servicebus/Azure.Messaging.ServiceBus/src/Amqp/AmqpConsumer.cs]: nit: In other methods we are checking AssertNotClosed `ServiceBusReceiver`. Can we update the same here?

File: sdk/servicebus/Azure.Messaging.ServiceBus/src/Amqp/AmqpConsumer.cs [hunk 11]
```diff
@@ -425,46 +1030,108 @@ internal List<ArraySegment<byte>> ConvertLockTokensToDeliveryTags(IEnumerable<Gu
...
-            var responseAmqpMessage = await Task.Factory.FromAsync(
-                (c, s) => requestResponseAmqpLink.BeginRequest(amqpMessage, transactionId, timeout, c, s),
-                (a) => requestResponseAmqpLink.EndRequest(a),
-                this).ConfigureAwait(false);
-
-            return AmqpResponseMessage.CreateResponse(responseAmqpMessage);
+            return messages;
         }
 
+
```
Comment (ShivangiReja) [inline sdk/servicebus/Azure.Messaging.ServiceBus/src/Amqp/AmqpConsumer.cs]: nit: extra line

File: sdk/servicebus/Azure.Messaging.ServiceBus/src/Amqp/AmqpConsumer.cs (hunk 5 above)
Comment (JoshLove-msft) [reply sdk/servicebus/Azure.Messaging.ServiceBus/src/Amqp/AmqpConsumer.cs]: Removed

File: sdk/servicebus/Azure.Messaging.ServiceBus/src/Amqp/AmqpConsumer.cs (hunk 6 above)
Comment (JoshLove-msft) [reply sdk/servicebus/Azure.Messaging.ServiceBus/src/Amqp/AmqpConsumer.cs]: Updated

File: sdk/servicebus/Azure.Messaging.ServiceBus/src/Amqp/AmqpConsumer.cs (hunk 9 above)
Comment (JoshLove-msft) [reply sdk/servicebus/Azure.Messaging.ServiceBus/src/Amqp/AmqpConsumer.cs]: Should be Ilist for return types but IEnumerable for input types.

File: sdk/servicebus/Azure.Messaging.ServiceBus/src/Amqp/AmqpConsumer.cs (hunk 10 above)
Comment (JoshLove-msft) [reply sdk/servicebus/Azure.Messaging.ServiceBus/src/Amqp/AmqpConsumer.cs]: Updated

File: sdk/servicebus/Azure.Messaging.ServiceBus/src/Amqp/AmqpConsumer.cs (hunk 6 above)
Comment (SeanFeldman) [reply sdk/servicebus/Azure.Messaging.ServiceBus/src/Amqp/AmqpConsumer.cs]: With the silly `Async` would be dropped altogether. Or the synchronous methods suffixed with `Sync` 😃

File: sdk/servicebus/Azure.Messaging.ServiceBus/src/Amqp/AmqpReceiver.cs [hunk 12]
```diff
@@ -337,20 +395,370 @@ private void CloseLink(ReceivingAmqpLink link)
...
+        /// Deferring messages does not impact message's expiration, meaning that deferred messages can still expire.
+        /// This operation can only be performed on messages that were received by this receiver.
+        /// </remarks>
+        public override async Task DeferAsync(
+            ServiceBusReceivedMessage message,
+            IDictionary<string, object> propertiesToModify = null,
+            CancellationToken cancellationToken = default)
+        {
+            Argument.AssertNotClosed(IsClosed, nameof(ServiceBusReceiver));
+            cancellationToken.ThrowIfCancellationRequested<TaskCanceledException>();
```
Comment (SeanFeldman) [inline sdk/servicebus/Azure.Messaging.ServiceBus/src/Amqp/AmqpReceiver.cs]: Is anything other than `TaskCanceledException` that will be thrown by this extension method?
